src/sorted_set.h

```diff
@@ -0,0 +1,94 @@
+#ifndef SORTED_SET_H
+#define SORTED_SET_H
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+
+/// Modifying a key in associative containers leads to UB. This is an std::set
+/// reimplementation with worse characteristics (such as re-sorting the
+/// container on every push), but one that allows for elements to be
+/// out-of-order (which can happen after a key was modified)
+template<class Elem, class Compare>
+struct SortedSet {
+  SortedSet(): comp(Compare()) {}
+  size_t size() const {
+    return data.size();
+  }
+
+  bool has_element(const Elem& elem) const {
+    return std::find(data.begin(), data.end(), elem) != data.end();
```

Comment:
[`std::find`](https://en.cppreference.com/w/cpp/algorithm/find) is O(n). You want [`std::binary_search`](https://en.cppreference.com/w/cpp/algorithm/binary_search) here.